src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/TempPE/DesignTimeInputsChangeTracker.cs

```diff
@@ -0,0 +1,238 @@
+
+            _broadcastBlock = DataflowBlockSlim.CreateBroadcastBlock<IProjectVersionedValue<DesignTimeInputsDelta>>(nameFormat: nameof(DesignTimeInputsChangeTracker) + "Broadcast {1}");
+            _publicBlock = AllowSourceBlockCompletion ? _broadcastBlock : _broadcastBlock.SafePublicize();
+
+            IDisposable projectLink = ProjectDataSources.SyncLinkTo(
+                   _inputsDataSource.SourceBlock.SyncLinkOptions(
+                       linkOptions: DataflowOption.PropagateCompletion),
+                   _projectSubscriptionService.ProjectRuleSource.SourceBlock.SyncLinkOptions(
+                      linkOptions: DataflowOption.WithRuleNames(ConfigurationGeneral.SchemaName)),
+                   _inputsActionBlock,
+                   DataflowOption.PropagateCompletion,
+                   cancellationToken: _project.Services.ProjectAsynchronousTasks.UnloadCancellationToken);
+
+            // Create an action block to process file change notifications
+            _fileWatcherActionBlock = DataflowBlockSlim.CreateActionBlock<IProjectVersionedValue<string[]>>(ProcessFileChangeNotification);
+            IDisposable watcherLink = _fileWatcher.SourceBlock.LinkTo(_fileWatcherActionBlock, DataflowOption.PropagateCompletion);
+
+            _disposables.AddDisposable(projectLink);
+            _disposables.AddDisposable(watcherLink);
+
+            JoinUpstreamDataSources(_inputsDataSource, _projectSubscriptionService.ProjectRuleSource, _fileWatcher);
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            if (_inputsActionBlock != null)
+            {
+                // This will stop our blocks taking any more input
+                _inputsActionBlock.Complete();
+                _fileWatcherActionBlock!.Complete();
+
+                _threadingService.ExecuteSynchronously(() => Task.WhenAll(_inputsActionBlock.Completion, _fileWatcherActionBlock.Completion));
+            }
+
+            _disposables.Dispose();
+        }
+
+        internal void ProcessFileChangeNotification(IProjectVersionedValue<string[]> arg)
+        {
+            // Ignore any file changes until we've received the first set of design time inputs (which shouldn't happen anyway)
+            // That first update will send out all of the files so we're not losing anything
+            if (_latestDesignTimeInputs == null)
+            {
+                return;
+            }
+
+            // Make sure the design time inputs don't change while we're processing this notification
+            DesignTimeInputs designTimeInputs = _latestDesignTimeInputs;
+
+            var changedFiles = new List<DesignTimeInputFileChange>();
+            foreach (string changedFile in arg.Value)
+            {
+                string relativeFilePath = _project.MakeRelative(changedFile);
+
+                // if a shared input changes, we recompile everything
+                if (designTimeInputs.SharedInputs.Contains(relativeFilePath))
```

Comment:
It's not a real problem because all of the inputs come from the same source (the projects items), but I was going to change these to be HashSets in a future PR anyway, I guess I'll just do that now.

Reply:
Wow I was putting off that change, but this new design makes it so simple :)